worker/src/main/scala/org/apache/celeborn/service/deploy/worker/monitor/JVMQuake.scala

```diff
@@ -0,0 +1,201 @@
+
+  private[this] val enabled = conf.workerJvmQuakeEnabled
+  private[this] val checkInterval = conf.workerJvmQuakeCheckInterval
+  private[this] val runtimeWeight = conf.workerJvmQuakeRuntimeWeight
+  private[this] val dumpThreshold = conf.workerJvmQuakeDumpThreshold.toNanos
+  private[this] val killThreshold = conf.workerJvmQuakeKillThreshold.toNanos
+  private[this] val dumpEnabled = conf.workerJvmQuakeDumpEnabled
+  private[this] val dumpPath = conf.workerJvmQuakeDumpPath
+  private[this] val exitCode = conf.workerJvmQuakeExitCode
+  private[this] val dumpDir =
+    s"${Utils.createDirectory(FileUtils.getTempDirectoryPath).getPath}/jvm-quake/dump"
+
+  private[this] var lastExitTime: Long = 0L
+  private[this] var lastGCTime: Long = 0L
+  private[this] var bucket: Long = 0L
+  private[this] var scheduler: ScheduledExecutorService = _
+
+  def start(): Unit = {
+    if (enabled) {
+      lastExitTime = getLastExitTime
+      lastGCTime = getLastGCTime
+      scheduler = ThreadUtils.newDaemonSingleThreadScheduledExecutor("jvm-quake")
+      scheduler.scheduleWithFixedDelay(
+        new Runnable() {
+          override def run(): Unit = {
+            JVMQuake.this.run()
+          }
+        },
+        0,
+        checkInterval,
+        TimeUnit.MILLISECONDS)
+    }
+  }
+
+  def stop(): Unit = {
+    if (enabled) {
+      scheduler.shutdown()
+    }
+  }
+
+  private def run(): Unit = {
+    val currentExitTime = getLastExitTime
+    val currentGCTime = getLastGCTime
+    val gcTime = currentGCTime - lastGCTime
+    val runTime = currentExitTime - lastExitTime - gcTime
+
+    bucket = Math.max(0, bucket + gcTime - BigDecimal(runTime * runtimeWeight).toLong)
+    logDebug(s"Time: (gc time: $gcTime, execution time: $runTime)")
+    logDebug(
+      s"Capacity: (bucket: $bucket,  dump threshold: $dumpThreshold, kill threshold: $killThreshold)")
+
+    if (bucket > dumpThreshold) {
+      logError(s"JVM GC has reached the threshold: bucket: $bucket, dumpThreshold: $dumpThreshold.")
+      if (shouldHeapDump) {
+        val savePath = getHeapDumpSavePath
+        val linkPath = getHeapDumpLinkPath
+        heapDump(savePath, linkPath)
+      } else if (bucket > killThreshold) {
+        System.exit(exitCode)
+      }
+    }
+    lastExitTime = currentExitTime
+    lastGCTime = currentGCTime
+  }
+
+  def shouldHeapDump: Boolean = {
+    dumpEnabled && !heapDumped
+  }
+
+  @VisibleForTesting
+  def getHeapDumpSavePath: String =
+    dumpPath.getOrElse(s"$dumpDir/$getProcessId/hprof")
+
+  @VisibleForTesting
+  def getHeapDumpLinkPath: String =
+    s"$dumpDir/$getProcessId/link"
+
+  private def heapDump(savePath: String, linkPath: String, live: Boolean = false): Unit = {
+    val saveDir = new File(savePath)
+    if (!saveDir.exists()) {
+      saveDir.mkdirs()
+    }
+    val heapDump = new File(saveDir, dumpFile)
+    if (heapDump.exists()) {
+      logInfo(s"Heap dump of worker exists: $heapDump.")
+      heapDumped = true
```

Comment:
are we assuming that the worker repeatedly restarts and crashes, then only keep the first heap dump?

Reply:
@pan3793, yes.

Reply:
please escalate the log level to the warning and explain our assumption. in the future, we may want to make the behavior configurable, so that the user could decide the overwrite behavior